src/elasticsearch_cleanup/main.py

```diff
@@ -0,0 +1,204 @@
+"""This script is used for removing duplicate documents from OpenSearch indices.
+
+The main logic includes:
+1. Setting up the OpenSearch client.
+2. Fetching all documents from specified OpenSearch indices.
+3. Identifying duplicate documents based on comparison fields.
+4. Removing the duplicates if not in dry-run mode or logging the bulk actions if in dry-run mode.
+
+This scripts assumes:
+1. fields name doesn't contain any of ';', ':', ',', '.'.
+2. '_source' field is Elasticsearch is enabled.
+"""
+
+import json
+from itertools import tee
+from typing import Any, Iterator
+
+from opensearchpy import OpenSearch, helpers
+
+from elasticsearch_cleanup import config, consts
+from elasticsearch_cleanup.logger import get_logger
+from elasticsearch_cleanup.utils import (
+    get_value_from_dict,
+    parse_index_and_fields_pairs,
+    str_to_bool,
+)
+
+logger = get_logger(config.LOG_LEVEL)
+
+
+def remove_documents(
+    opensearch_client: OpenSearch,
+    actions: list[dict[str, str]],
+    index: str,
+):
+    """Removes specified documents from the OpenSearch index.
+
+    Args:
+        opensearch_client: The client to communicate with OpenSearch.
+        actions: List of actions to remove documents.
+        index: The name of the index.
+    """
+    bulk_errors = helpers.bulk(client=opensearch_client, actions=actions)
+    logger.info(f"Removed '{len(actions)}' duplicates from '{index}'.")
+
+    opensearch_client.indices.refresh(index=index)
+
+    if len(bulk_errors) > 0:
+        raise ExceptionGroup("failed to bulk some actions", bulk_errors)  # type: ignore
+
+
+def get_bulk_actions(
+    documents: Iterator,
+    comparison_fields: list[str],
+    index: str,
+) -> list[dict[str, str]]:
+    """Generates bulk actions for documents based on a comparison field.
+
+    Args:
+        documents: List of documents to be compared.
+        comparison_fields: List of fields for comparison to identify unique documents.
+        index: The name of the index.
+
+    Returns:
+        List of bulk actions for the documents.
+    """
+    seen_values = set()
+    bulk_actions = []
+
+    for doc in documents:
+        unique_fields_value = tuple(
+            get_value_from_dict(
+                dot_notation_string=field,
+                data=doc["_source"],
+            )
+            for field in comparison_fields
+        )
+
+        if unique_fields_value in seen_values:
+            bulk_actions.append(
+                {
+                    "_op_type": "delete",
+                    "_index": index,
+                    "_id": doc["_id"],
+                }
+            )
+        else:
+            seen_values.add(unique_fields_value)
+
+    return bulk_actions
+
+
+def fetch_all_documents(
```

Comment:
just inline the function. only one occurrence of it and documentation is longer than implementation